velox/vector/DecodedVector.cpp

```diff
@@ -487,41 +480,32 @@ DecodedVector::DictionaryWrapping DecodedVector::dictionaryWrapping(
     vector_size_t size) const {
   VELOX_CHECK(!isIdentityMapping_);
   VELOX_CHECK(!isConstantMapping_);
-
   VELOX_CHECK_LE(size, size_);
 
-  if (isOneLevelDictionary(wrapper)) {
-    // Re-use indices and nulls buffers.
-    return {wrapper.wrapInfo(), wrapper.nulls()};
-  } else {
-    // Make a copy of the indices and nulls buffers.
-    BufferPtr indices = copyIndicesBuffer(indices_, size, wrapper.pool());
-    // Only copy nulls if we have nulls coming from one of the wrappers, don't
-    // do it if nulls are missing or from the base vector.
-    BufferPtr nulls = hasExtraNulls_
-        ? copyNullsBuffer(nulls_, size, wrapper.pool())
-        : nullptr;
-    return {std::move(indices), std::move(nulls)};
-  }
+  // Make a copy of the indices and nulls buffers.
+  BufferPtr indices = copyIndicesBuffer(indices_, size, wrapper.pool());
+  // Only copy nulls if we have nulls coming from one of the wrappers, don't
+  // do it if nulls are missing or from the base vector.
+  // TODO: remove the check for hasExtraNulls_ after #3553 is merged.
+  BufferPtr nulls =
+      hasExtraNulls_ ? copyNullsBuffer(nulls_, size, wrapper.pool()) : nullptr;
+  return {std::move(indices), std::move(nulls)};
 }
 
 VectorPtr DecodedVector::wrap(
     VectorPtr data,
     const BaseVector& wrapper,
     vector_size_t size) {
-  // Return `data` as is if it is constant encoded and the vector size matches
-  // exactly with the selection size. Otherwise, the constant vector will need
-  // to be resized to match it. The resizing will be done in the wrapping code
-  // later.
-  if (data->isConstantEncoding() && size == data->size()) {
-    return data;
-  }
-
-  if (wrapper.isConstantEncoding()) {
-    if (wrapper.isNullAt(0)) {
+  if (isConstantMapping_) {
+    if (isNullAt(0)) {
       return BaseVector::createNullConstant(data->type(), size, data->pool());
+    } else if (data->isConstantEncoding() && size == data->size()) {
+      // Return `data` as is if it is constant encoded and the vector size
+      // matches exactly with the selection size. Otherwise, the constant vector
+      // will need to be resized to match it.
+      return data;
     }
-    return BaseVector::wrapInConstant(size, wrapper.wrappedIndex(0), data);
+    return BaseVector::wrapInConstant(size, constantIndex_, data);
```

Comment:
The comment in DecodedVector.h says "Applies only when isConstantMapping_ is true and baseVector_ is of complex type". This function applies to scalar vectors too, right? Would this be a concern?

Reply:
nit: Maybe we could add a comment saying "If wrapper is constant mapping, data is expected to be constant as well". Although for complex-typed `data` this is not required, I feel it's still a reasonable assumption based on how this API is used in peeling logic.

Reply:
@kagamiori
I am not really sure if there is a terse way of convening this without the whole context about using this for peeling.

Should I instead add context about this to the class comment where I can be a bit more verbose. something like:
“If after decoding, the wrap and base both are constant then calling wrap() on a non-constant vector would result in unpredictable behavior. This is because the base() in this case can only be a scalar constant which renders the constant index in the wrap irrelevant.”